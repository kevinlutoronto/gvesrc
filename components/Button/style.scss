@import '../../styles/base';

.button {
  @extend %button--base;
  @extend %type--body;
}

.primary {
  @extend %button--primary;
}

.secondary {
  @extend %button--secondary;
}

.tertiary {
  @extend %button--tertiary;
}

.loading {
  opacity: 0.9;
  cursor: default;
  pointer-events: none;
}

.small {
  @extend %button--small;
}
