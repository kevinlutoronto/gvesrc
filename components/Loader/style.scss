@import '../../styles/base';

.container {
  width: 100%;
  height: 100%;
  margin: 0 auto;
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: auto;
}

.spinner {
  margin: auto 0;
  text-align: center;
}

.spinner > div {
  width: 8px;
  height: 8px;
  background-color: #fff;
  border-radius: 100%;
  display: inline-block;
  margin: 0 2px;
  -webkit-animation: sk-bouncedelay 1.4s infinite ease-in-out both;
  animation: sk-bouncedelay 1.4s infinite ease-in-out both;
}

.spinner .bounce1 {
  -webkit-animation-delay: -0.32s;
  animation-delay: -0.32s;
  background-color: $cPurple200;
}

.spinner .bounce2 {
  -webkit-animation-delay: -0.16s;
  animation-delay: -0.16s;
  background-color: $cPurple200;
}

.spinner .bounce3 {
  background-color: $cPurple200;
}

@-webkit-keyframes sk-bouncedelay {
  0%,
  80%,
  100% {
    -webkit-transform: scale(0);
  }
  40% {
    -webkit-transform: scale(1);
  }
}

@keyframes sk-bouncedelay {
  0%,
  80%,
  100% {
    -webkit-transform: scale(0);
    transform: scale(0);
  }
  40% {
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}
