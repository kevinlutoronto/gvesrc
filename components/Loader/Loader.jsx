// @flow
import React from 'react';
import cn from 'classnames';
import style from './style.scss';
import { getDataAttrs } from '../../dataUtils';
import type { Data, StyleAttributes } from '../../types';

type Props = {|
  className?: string,
  dotColor?: string,
  dotClassName?: string,
  style?: StyleAttributes,
  dotStyle?: StyleAttributes,
  data?: Data,
|};

class Loader extends React.Component<Props> {
  static defaultProps = {
    dotStyle: {},
  };

  render() {
    const { dotColor, dotStyle = {} } = this.props;
    if (dotColor != null) {
      dotStyle.backgroundColor = dotColor;
    }

    return (
      <div
        {...getDataAttrs(this.props.data)}
        className={cn(style.container, this.props.className)}
        style={this.props.style}
      >
        <div className={style.spinner}>
          <div className={cn(style.bounce2, this.props.dotClassName)} style={dotStyle} />
          <div className={cn(style.bounce3, this.props.dotClassName)} style={dotStyle} />
          <div className={cn(style.bounce1, this.props.dotClassName)} style={dotStyle} />
        </div>
      </div>
    );
  }
}

Loader.displayName = 'Plasma@Loader';

export default Loader;
