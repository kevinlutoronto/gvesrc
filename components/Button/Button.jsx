// @flow
import cx from 'classnames';
import React from 'react';

import Loader from '../Loader/Loader';

import { getDataAttrs } from '../../dataUtils';

import type { Data, StyleAttributes } from '../../types';

import styles from './style.scss';

const Variants = Object.freeze({
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
});
const Sizes = Object.freeze({
  SMALL: 'small',
});

type Props = {|
  autofocus?: boolean,
  children?: React$Node,
  className?: string,
  data?: Data,
  disabled?: boolean,
  hidden?: boolean,
  label?: string,
  loading?: boolean,
  name?: string,
  onClick?: (evt: MouseEvent) => void,
  id?: string,
  isSubmit?: boolean,
  size?: $Values<typeof Sizes>,
  style?: StyleAttributes,
  type?: $Values<typeof Variants>,
  value?: string,
  tabIndex?: number,
  title?: string,
|};

function Button(props: Props): React$Node {
  const {
    children,
    className,
    data,
    disabled = false,
    isSubmit,
    label,
    loading = false,
    onClick,
    size,
    style,
    type = Variants.PRIMARY,
    // eslint-disable-next-line comma-dangle
    ...rest
  } = props;

  const classes = cx(styles.button, className, {
    [styles.primary]: type === Variants.PRIMARY,
    [styles.secondary]: type === Variants.SECONDARY,
    [styles.tertiary]: type === Variants.TERTIARY,
    [styles.small]: size === Sizes.SMALL,
    [styles.loading]: loading,
  });

  let contentComponent;
  if (loading) {
    const loaderDotStyle =
      type === Variants.SECONDARY || type === Variants.TERTIARY
        ? { backgroundColor: '#000', opacity: '0.1' }
        : undefined;

    contentComponent = <Loader dotStyle={loaderDotStyle} />;
  } else {
    contentComponent = children || label;
  }

  return (
    // eslint-disable-next-line react/button-has-type
    <button
      {...getDataAttrs(data)}
      className={classes}
      disabled={disabled}
      onClick={onClick}
      style={style}
      type={isSubmit ? 'submit' : 'button'}
      {...rest}
    >
      {contentComponent}
    </button>
  );
}

Button.displayName = 'Plasma@Button';

export default Button;
