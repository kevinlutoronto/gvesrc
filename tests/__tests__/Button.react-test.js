/* eslint-disable */

import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

import Button from '../../components/Button/Button';
import Loader from '../../components/Loader/Loader';

describe('Button Component', () => {
  test('render', () => {
    const component = renderer.create(<Button>YOLO</Button>);
    expect(component).toMatchSnapshot();
  });

  test('children', () => {
    const component = shallow(<Button>YOLO</Button>);
    expect(component.contains('YOLO')).toBe(true);
  });

  test('primary', () => {
    const component = shallow(<Button>YOLO</Button>);
    expect(component.find('.primary')).toHaveLength(1);
    expect(component.find('.secondary')).toHaveLength(0);
    expect(component.find('.tertiary')).toHaveLength(0);
  });

  test('secondary', () => {
    const component = shallow(<Button type="secondary">YOLO</Button>);
    expect(component.find('.primary')).toHaveLength(0);
    expect(component.find('.secondary')).toHaveLength(1);
    expect(component.find('.tertiary')).toHaveLength(0);
  });

  test('tertiary', () => {
    const component = shallow(<Button type="tertiary">YOLO</Button>);
    expect(component.find('.primary')).toHaveLength(0);
    expect(component.find('.secondary')).toHaveLength(0);
    expect(component.find('.tertiary')).toHaveLength(1);
  });

  test('disabled', () => {
    const component = shallow(<Button disabled>YOLO</Button>);
    expect(component.find('[disabled]')).toHaveLength(1);
    expect(component.prop('disabled')).toBe(true);
  });

  test('small', () => {
    const component = shallow(<Button size="small">YOLO</Button>);
    expect(component.find('.small')).toHaveLength(1);
  });

  test('loading', () => {
    const component = shallow(<Button loading>YOLO</Button>);
    expect(component.find('.loading')).toHaveLength(1);
    expect(component.contains('YOLO')).toBe(false);
    expect(component.find(Loader)).toHaveLength(1);
  });

  test('submit', () => {
    const component = shallow(<Button isSubmit>YOLO</Button>);
    expect(component.prop('type')).toBe('submit');
  });

  test('onClick', () => {
    const onClick = jest.fn();
    const component = mount(<Button onClick={onClick}>YOLO</Button>);
    component.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('onClick disabled', () => {
    const onClick = jest.fn();
    const component = mount(
      <Button disabled onClick={onClick}>
        YOLO
      </Button>
    );
    component.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(0);
  });

  test('with data attribute', () => {
    const component = shallow(<Button data={{ qa: 'important' }}>YOLO</Button>);
    expect(component.prop('data-qa')).toBe('important');
  });
});
