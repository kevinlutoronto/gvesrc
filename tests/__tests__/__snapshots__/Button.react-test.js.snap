// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component render 1`] = `
<button
  className="button primary"
  disabled={false}
  onClick={undefined}
  style={undefined}
  type="button"
>
  YOLO
</button>
`;
